clients/src/main/java/org/apache/kafka/clients/consumer/internals/metrics/KafkaAsyncConsumerMetrics.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.consumer.internals.metrics;
+
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Value;
+
+import java.util.Arrays;
+
+import static org.apache.kafka.clients.consumer.internals.ConsumerUtils.CONSUMER_METRICS_SUFFIX;
+
+public class KafkaAsyncConsumerMetrics extends KafkaConsumerMetrics implements AutoCloseable {
```

Comment:
These metrics are really to do with measuring the processing of events sent between the application and background threads. AsyncKafkaConsumer and ShareConsumerImpl both use the same architecture and the same metrics would apply equally well to both. So, I suggest:

* Renaming this to `AsyncConsumerMetrics`
* Do not inherit from `KafkaConsumerMetrics`
* Supply an instance of this class when constructing the `NetworkClientDelegate` in `ShareConsumerImpl`. This should not be `Optional.empty()`. In fact, I don't think it should be optional at all.

Reply:
Hi @AndrewJSchofield, thanks for the suggestion. I rename it to `AsyncConsumerMetrics` and remove `Optional.empty()`. However, it still needs to inherit from `KafkaConsumerMetrics`, because `AsyncKafkaConsumer` implements `ConsumerDelegate` which returns `KafkaConsumerMetrics` in `kafkaConsumerMetrics` function.

https://github.com/apache/kafka/blob/cebec91470c440a1735e50210ae3fc0f62ee40a7/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L2079-L2082

https://github.com/apache/kafka/blob/cebec91470c440a1735e50210ae3fc0f62ee40a7/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegate.java#L40

Reply:
Alternatively, you could do a bit more refactoring. There are essentially two metrics objects now - the KafkaConsumerMetrics and the AsyncConsumerMetrics. There could be two methods in `ConsumerDelegate` to give access to them.

`ShareConsumerDelegate` could follow the same pattern.

What do you think? I feel there's a bit more tidying up to do in this area.

Reply:
Hi @AndrewJSchofield, thanks for the suggestion. If we want to do this change, we also need to add a new function `asyncConsumerMetrics` to `KafkaConsumer` like `KafkaConsumer#kafkaConsumerMetrics`. Before I update the PR, I would like to confirm we are on the same page. Thanks.

Reply:
Hey folks, do you want to address this in this PR or do the refactor in a follow-up? Thoughts @AndrewJSchofield?